### PR TITLE
JCLOUDS-1216 - oneandone-user-api

### PR DESCRIPTION
**Pull request by alibazlamit**
@nacx the Users PR is ready for review, i had one issue with this specific PR since the API does not allow to use many of the endpoints that affect the user info unless its the same user that is being used to make the request. all these tests are commented out now.

### CODE REVIEW

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/User.java [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+   public abstract String name();
+
+   @Nullable
+   public abstract String description();
+
+   @Nullable
+   public abstract String email();
+
+   @Nullable
+   public abstract String state();
```
Comment (nacx) [inline oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/User.java]: Should this be of type `UserState`?

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/User.java [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+   }
+
+   @AutoValue
+   public abstract static class AddUserIp {
+
+      public abstract List<String> ips();
+
+      @SerializedNames({"ips"})
+      public static AddUserIp create(List<String> ips) {
+         return new AutoValue_User_AddUserIp(ips == null ? ImmutableList.<String>of() : ips);
```
Comment (nacx) [inline oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/User.java]: Enforce an immutable list if the parameter is not null.

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/User.java [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+      }
+   }
+
+   @AutoValue
+   public abstract static class Api {
+
+      public abstract Boolean active();
```
Comment (nacx) [inline oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/User.java]: Prefer primitive types for non-nullable fields.

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/User.java [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+
+      @Nullable
+      public abstract String key();
+
+      @Nullable
+      public abstract List<String> allowedIps();
+
+      @SerializedNames({"active", "key", "allowed_ips"})
+      public static Api create(Boolean active, String key, List<String> allowedIps) {
+         return new AutoValue_User_Api(active, key, allowedIps == null ? ImmutableList.<String>of() : allowedIps);
```
Comment (nacx) [inline oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/User.java]: Immutable when not null.

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/User.java [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+      }
+
+      @AutoValue
+      public abstract static class UpdateApi {
+
+         public abstract Boolean active();
```
Comment (nacx) [inline oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/User.java]: Same about primitives.

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/UserApi.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+import org.jclouds.Fallbacks;
+import org.jclouds.rest.annotations.BinderParam;
+import org.jclouds.rest.annotations.Fallback;
+import org.jclouds.rest.annotations.MapBinder;
+import org.jclouds.rest.annotations.RequestFilters;
+import org.jclouds.rest.annotations.SelectJson;
+import org.jclouds.rest.binders.BindToJsonPayload;
+
+@Path("/users")
+@Produces("application/json")
```
Comment (nacx) [inline oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/UserApi.java]: Remove this and leave only the method specific ones

File: oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/UserApiLiveTest.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+   @Test(dependsOnMethods = "testGet")
+   public void testUpdate() throws InterruptedException {
+      //this test will fail too since its not allowed to update other users from the API
+//      String updatedDescription = "updatejclouds";
+//
+//      User updateResult = userApi().update(currentUser.id(), User.UpdateUser.builder()
+//              .description(updatedDescription)
+//              .build());
+//
+//      assertEquals(updateResult.description(), updatedDescription);
```
Comment (nacx) [inline oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/UserApiLiveTest.java]: Can't we configure these tests to update the logged user? Could create a new API context for the just created user and use that to implement these tests?

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/UserApi.java (hunk 6 above)
Comment (alibazlamit) [inline oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/UserApi.java]: `@Produces("application/json")` we have been using this approach so far not a method specific tag, you recommended this approach in the beginning.

Comment (nacx) [inline oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/UserApi.java]: Hmmm I think there was some misunderstanding then. We should only add the "content-type" headers in the requests that generate a body, and that's what that annotation does. If we put it at class level, all requests will have it, even GET requests without body. I don't know if the API needs it, but we'd better put it just where needed.

File: oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/UserApiLiveTest.java (hunk 7 above)
Comment (nacx) [inline oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/UserApiLiveTest.java]: Hmmm I don't really know. The questions to answer are:

* Why would we have tests we won't be able to run?
* Under which circumstances could we run a complete test suite for this API? What preconditions should be met?
* Can we dynamically configure those preconditions so we can run the test suite?

Comment (alibazlamit) [inline oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/UserApiLiveTest.java]: @nacx 
I think that this PR and the Users API does not make a lot of sense to jClouds, most of the operations are only allowed to the current user, and changing the API key or deleting the users might break any flow, if you agree i think this feature could be skipped for jClouds, WDYT?

Comment (nacx) [inline oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/UserApiLiveTest.java]: I agree this should not be our highest priority.
Moving forward, do you think the existing APIs cover the funcionality to implement the ComputeService abstraction?

**Comment by alibazlamit** [comment]
I do believe so, we have covered all the important server operations, I will delete this PR and start a new one for the compute api.

File: oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/UserApiLiveTest.java (hunk 7 above)
Comment (alibazlamit) [inline oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/UserApiLiveTest.java]: I do believe so, we have covered all the important server operations, I will delete this PR and start a new one for the compute api.

Comment (nacx) [inline oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/UserApiLiveTest.java]: That sounds great. Thanks! :)

**Pull request closed without merging**